### PR TITLE
Remove transform

### LINKED ISSUE #12674: Remove mapping source transform scripts
The ability to transform the `_source` field before indexing (while storing the original `_source`) is confusing and error prone.  Most of what you can do with it can already be done with `copy_to`.  It also makes it difficult to know where a field value has come from, which means it can't support highlighting.

I think this is not widely used at all and should be removed in 2.0 (while continuing to support indices which already use it for 2.x releases)

### PR DESCRIPTION
**Pull request by nik9000**
Removes the mapping transform feature which when used made debugging very
difficult. Users should transform their documents on the way into
Elasticsearch rather than having Elasticsearch do it.

Closes #12674

### CODE REVIEW

**Comment by nik9000** [comment]
Ping @rjernst for review.

**Comment by jpountz** [comment]
The change looks good to me, maybe we should document it (unless it already is?) It's a bit weird to me to drop a feature on a minor release. Is it feasible to either deprecate for now and only remove in 3.0 or backport the change to 2.0 since it's not been released yet? cc @clintongormley

File: core/src/test/java/org/elasticsearch/script/expression/ExpressionScriptIT.java [hunk 1]
```diff
@@ -493,32 +493,6 @@ public void testInvalidUpdateScript() throws Exception {
         }
     }
 
-    // test to make sure expressions are not allowed to be used as mapping scripts
-    public void testInvalidMappingScript() throws Exception{
```
Comment (rjernst) [inline core/src/test/java/org/elasticsearch/script/expression/ExpressionScriptIT.java]: I think the entire "mapping" `ScriptContext` can be removed?

File: core/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java [hunk 2]
```diff
@@ -126,24 +123,9 @@ public Builder put(MetadataFieldMapper.Builder<?, ?> mapper) {
             return this;
         }
 
-        public Builder transform(ScriptService scriptService, Script script) {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java]: `ScriptService` can be removed from `MapperService` right?

**Comment by rjernst** [comment]
Also LGTM (just a couple extra things I think can be removed).

I think 3.0 is the correct place to put this, since we missed the window to get this into 2.0. Heaver mappings cleanup will just have to wait a little longer. :)  But let's make sure to deprecate in 2.1.

**Comment by nik9000** [comment]
> I think 3.0 is the correct place to put this, since we missed the window to get this into 2.0. Heaver mappings cleanup will just have to wait a little longer. :) But let's make sure to deprecate in 2.1.

We blasted the documentation in 2.0, IIRC. I'm pretty ambivalent about when it gets removed but @clintongormley wanted to see it gone as soon as we could.

**Comment by clintongormley** [comment]
Question for me is: what features/improvements does mapping transform block?  eg highlighting?  are there things we want to get in for 2.x which are blocked?

**Comment by mattjanssen** [comment]
We just built and deployed a large platform on 2.0 that uses these transforms to extract a geo_point from a geo_shape. We need the geo_point to do distance sorting. We need the geo_shape to do  spatial queries.

We assumed it was a safe feature because it hadn't been deprecated. We didn't even notice it wasn't in the latest docs.

We use mongo-connector to ship the data. We don't want to store the lon/lat twice in our DB just to facilitate Elasticsearch. I guess we should rewrite mongo-connector to do transforms?

**Comment by lukas-vlcek** [comment]
Hmm... I liked mapping transformations.

**Comment by jpountz** [comment]
Maybe node ingest (#14047) removes the need for transform? So we could remove in 3.0 as @rjernst suggested.

**Comment by nik9000** [comment]
> Maybe node ingest (#14047) removes the need for transform? So we could remove in 3.0 as @rjernst suggested.

Right then. Rebasing onto 3.0.

File: core/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java (hunk 2 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java]: Removing!

**Comment by nik9000** [comment]
Ok - @rjernst, I think this is ready for another round of review when you are ready for it.

**Comment by rjernst** [comment]
LGTM

**Comment by adichad** [comment]
I have been using mapping transform native (java) scripts all the way up to 1.7.x in production. never saw a deprecation warning in code or documentation online though. In 2.0 documentation the whole section is not present. the migration tool also did not warn about the transform script reference being present in the type mappings. 

Please provide better documentation than throwing a 404 in 2.x documentation at least stating whether this is deprecated or altogether removed in 2.x or not.

In the meanwhile I will look into 2.0 code to see whether I need to incur the effort to move this out now or whether it can wait a month or so.

**Comment by rjernst** [comment]
@adichad Mapping transforms still exist in 2.x. This PR is to remove them from master (3.0), and a follow up will deprecate them in 2.x.

**Comment by adichad** [comment]
Thanks @rjernst !
request to reinstate the related documentation for example at

https://www.elastic.co/guide/en/elasticsearch/reference/1.7/mapping-transform.html
https://www.elastic.co/guide/en/elasticsearch/reference/1.7/_get_transformed.html

to also work at

https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-transform.html
https://www.elastic.co/guide/en/elasticsearch/reference/current/_get_transformed.html

and maintain deprecation warnings there?

**Comment by nik9000** [comment]
Squashed, rebased, and merging to just 3.0.0.

**Comment by niemyjski** [comment]
Removing this feature really made life tough for us... Do you have any work around for this?

**Comment by niemyjski** [comment]
cc @nik9000 @rjernst @clintongormley

**Comment by niemyjski** [comment]
Don't see these changes in here: https://www.elastic.co/guide/en/elasticsearch/reference/5.x/breaking_50_mapping_changes.html We were using this feature like this: 
https://github.com/exceptionless/Exceptionless/blob/master/src/Exceptionless.Core/Repositories/Configuration/Indexes/EventIndex.cs#L191-L212

which now completely breaks us.

**Comment by jasontedor** [comment]
It's on the page that you linked to: https://www.elastic.co/guide/en/elasticsearch/reference/5.0/breaking_50_mapping_changes.html#_source_transform_removed

**Comment by niemyjski** [comment]
From what I understand there are issues with ingest pipelines around bulk operations??

**Comment by nik9000** [comment]
The idea is to replace it with the ingest feature. It isn't the same as the
modifications it makes to the source are saved in the source rather than
just used for indexing.

It was a neat feature but ended up breaking more folks than it helped
because it didn't work well with things like highlighting. It made
debugging indexing problems a nightmare.

On Oct 11, 2016 9:15 PM, "Jason Tedor" <notifications@github.com> wrote:

It's on the page that you linked to: https://www.elastic.co/guide/
en/elasticsearch/reference/5.0/breaking_50_mapping_changes.
html#_source_transform_removed

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/elastic/elasticsearch/pull/13657#issuecomment-253092240>,
or mute the thread

**Pull request merged**